### PR TITLE
Improve foreach reference

### LINKED ISSUE #4505: The unobvious implicit cast in foreach statements
Today, I've run against a quiet unobvious behavior of the C# compiler that I believe deserves to be documented. I've submitted an issue to the roslyn team (dotnet/roslyn#25080), but it looks like this behavior is considered correct (although I can't even find proper words to express how I disagree). 

There's a detailed description in the issue, in short: when the C# compiler has to process a statement like
```C#
foreach (T item in items)
{
}
```
where `items` is of type `IEnumerable<TImpl>` (or equivalent enumerable type as supported by `foreach`) and `T` is an `interface` that is not implemented by `TImpl`, it (compiler) adds a cast from `object` to `T` and allows the statement to compile. Thus, at some point at run time there might be an `InvalidCastException` upon an attempt to cast an `item` from `items` that does not implement `T`.

It think this needs to be documented somewhere and I haven't found any related docs. @mairaw what do you think about this?

edit by www: Add document details section so this displays in the issues control on the `foreach` reference page.

#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d710feeb-8f03-b211-a070-155938929d6b
* Version Independent ID: c980ea8e-ffb4-fbb1-1761-292858616cdd
* Content: [C# foreach statement](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/foreach-in)
* Content Source: [docs/csharp/language-reference/keywords/foreach-in.md](https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/keywords/foreach-in.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #4505 

Fixes [AB#1702034](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1702034)

First three commits move all snippets to the new correct location.
Fourth commit adds new language for implicit cast of the iteration variable.
Final commit is Acrolinx feedback.

### CODE REVIEW

**Review by IEvangelist — DISMISSED** [review]
Oops, I reviewed just a single file -- didn't realize it was part of a larger draft... disregard for now.

**Comment by BillWagner** [comment]
It's now ready @IEvangelist

File: docs/csharp/programming-guide/generics/snippets/GenericWhereConstraints.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+        // <Snippet8>
+        delegate T MyDelegate<T>() where T : new();
+        // </Snippet8>
+    }
+
+    // <Snippet9>
+    public class Employee
+    {
+        public Employee(string s, int i) => (Name, ID) = (s, i);
```
Comment (IEvangelist) [inline docs/csharp/programming-guide/generics/snippets/GenericWhereConstraints.cs:79]:
```suggestion
        public Employee(string name, int id) => (Name, ID) = (name, id);
```

Prefer a bit more explicit parameter naming.

File: docs/csharp/programming-guide/generics/snippets/GenericWhereConstraints.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+            TestDelegateCombination();
+            TestEnumValues();
+        }
+
+        // <Snippet11>
+        public static void OpEqualsTest<T>(T s, T t) where T : class
+        {
+            System.Console.WriteLine(s == t);
+        }
+        private static void TestStringEquality()
```
Comment (IEvangelist) [inline docs/csharp/programming-guide/generics/snippets/GenericWhereConstraints.cs:217]:
```suggestion

        private static void TestStringEquality()
```

Add a line between methods.

File: docs/csharp/language-reference/keywords/foreach-in.md [hunk 3]
```diff
@@ -12,11 +12,23 @@ ms.assetid: 5a9c5ddc-5fd3-457a-9bb6-9abffcd874ec
...
 
 - has the public parameterless `GetEnumerator` method whose return type is either class, struct, or interface type,
 - the return type of the `GetEnumerator` method has the public `Current` property and the public parameterless `MoveNext` method whose return type is <xref:System.Boolean>.
 
+In most uses, `foreach` iterates an `IEnumerable<T>` expression, and the elements are the type `T`. However, the elements may be any type that has an implicit or explicit conversion from the type of the `Current` property. If the `Current` property returns `SomeType`, the type of the elements may be:
+
+- Any of the base classes of `SomeType`.
+- Any of the interfaces implemented by `SomeType`.
+
+Furthermore, if `SomeType` is a `class` or `interface` and not `sealed`, the type of elements may include:
```
Comment (IEvangelist) [inline docs/csharp/language-reference/keywords/foreach-in.md:25]:
```suggestion
Furthermore, if `SomeType` is a `class` or an `interface` and not `sealed`, the type of elements may include:
```

File: docs/csharp/language-reference/keywords/foreach-in.md [hunk 4]
```diff
@@ -12,11 +12,23 @@ ms.assetid: 5a9c5ddc-5fd3-457a-9bb6-9abffcd874ec
...
+The `foreach` statement executes a statement or a block of statements for each element in an instance of the type that implements the <xref:System.Collections.IEnumerable?displayProperty=nameWithType> or <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType> interface. The `foreach` statement isn't limited to those types and can be applied to an instance of any type that satisfies the following conditions:
 
 - has the public parameterless `GetEnumerator` method whose return type is either class, struct, or interface type,
 - the return type of the `GetEnumerator` method has the public `Current` property and the public parameterless `MoveNext` method whose return type is <xref:System.Boolean>.
 
+In most uses, `foreach` iterates an `IEnumerable<T>` expression, and the elements are the type `T`. However, the elements may be any type that has an implicit or explicit conversion from the type of the `Current` property. If the `Current` property returns `SomeType`, the type of the elements may be:
```
Comment (IEvangelist) [inline docs/csharp/language-reference/keywords/foreach-in.md:20]:
```suggestion
In most uses, `foreach` iterates an `IEnumerable<T>` expression, and where each element is of type `T`. However, the elements may be any type that has an implicit or explicit conversion from the type of the `Current` property. If the `Current` property returns `SomeType`, the type of the elements may be:
```

**Review by IEvangelist — APPROVED** [review]
Looks good to me, :shipit: when you're ready. A few minor suggestions, otherwise this looks great!

**Pull request merged**